src/hotspot/cpu/x86/c2_MacroAssembler_x86.cpp

```diff
@@ -1569,6 +1569,355 @@ void C2_MacroAssembler::vinsert(BasicType typ, XMMRegister dst, XMMRegister src,
   }
 }
 
+void C2_MacroAssembler::vpackI2X(BasicType elem_bt, XMMRegister dst,
+                                 XMMRegister ones, XMMRegister xtmp,
+                                 int vlen_enc) {
+  assert(VM_Version::supports_avx512vl(), "");
+  if (elem_bt == T_SHORT) {
+    evpmovdw(dst, dst, vlen_enc);
+  } else {
+    assert(elem_bt == T_BYTE, "");
+    evpmovdb(dst, dst, vlen_enc);
+  }
+}
+
+void C2_MacroAssembler::vgather_subword_avx3(BasicType elem_bt, XMMRegister dst, Register base, XMMRegister offset,
+                                             Register idx_base, int idx_off, XMMRegister idx_vec, XMMRegister ones,
+                                             XMMRegister xtmp, KRegister gmask, int vlen_enc) {
+  int mask_shift = -1;
+  int bitlevel_offset_shift = -1;
+  int nomlarized_index_shift = -1;
+  Address::ScaleFactor scale = Address::no_scale;
+
+  if (elem_bt == T_SHORT) {
+    mask_shift = 31;
+    scale = Address::times_2;
+    bitlevel_offset_shift = 4;
+    nomlarized_index_shift = 1;
+  } else {
+    assert(elem_bt == T_BYTE, "");
+    mask_shift = 30;
+    scale = Address::times_1;
+    bitlevel_offset_shift = 3;
+    nomlarized_index_shift = 2;
+  }
+
+  evmovdquq(idx_vec, Address(idx_base, idx_off, Address::times_4), vlen_enc);
+  if (offset != xnoreg) {
+    vpaddd(idx_vec, idx_vec, offset, vlen_enc);
+  }
+  // Normalize the indices to multiple of nomlarized_index_shift.
+  vpslld(xtmp, ones, nomlarized_index_shift, vlen_enc);
+  vpand(xtmp, idx_vec, xtmp, vlen_enc);
+  // Load double words from normalized indices.
+  evpgatherdd(dst, gmask, Address(base, xtmp, scale), vlen_enc);
```

Comment:
Another question, looks to me that we could read beyond the allocated memory for the array here. e.g. consider the following case:
* It is a byte gather
* The byte source array is of size 41, i.e. only indices 0-40 are valid
* The gather index is 40

Then as part of evpgatherdd we would be reading bytes at 40-43 offset from source array.

Reply:
I guess the fact that the Java objects are 8 byte alignment padded and the alignment being done at lines 1609-1611 and 1616-1621 somehow takes care of this.

Reply:
Drive by comment, that can change with Project Lilliput.

Reply:
Exactly!!, edge case may overrun with double word gather solution, this was not my original proposal, rolling back looks reasonable for the time being. WDYT ?

Reply:
I don't think Lilliput will relax the array alignment to below 4 bytes since the array length needs to be aligned at that.

Reply:
Hi @sviswa7 , I have rolled back to originally proposed solution.

Reply:
> I don't think Lilliput will relax the array alignment to below 4 bytes since the array length needs to be aligned at that.

HI @merykitty 

Lilliput wiki mention about shrinking object header below 4 bytes in future. 
https://wiki.openjdk.org/display/lilliput#:~:text=Future%20work%3A%2032%20bit%20or%20even%20smaller%20header%3F%20(If%20so%3A%20improve%20field%20layout%20to%20put%20fields%20in%20unused%20bits%20of%20header%20word)

On x86 page size always a power of 2, this along with minimum object size alignment in Java of 8 bytes even 4 bytes should be ok, We are anyway shuffling and extracting the byte / word of interest so any junk data read will be chopped off. But in edge case an overrun may land us into a non accessible address space, so we want to be super safe here. I am also sitting over other solution which is performant.

Reply:
Hi @jatin-bhateja,

The layout of an array is as follows:

    [header] - [length] - [data]

Since `length` is a 4-byte field, which is 4-byte aligned, in the foreseeable future, even if the `header` is shrunk to below 4 bytes, I don't think `data` can be less than 4-byte aligned.

Thanks a lot.

Reply:
Hi @merykitty , 

As mentioned earlier, JVM commits heap space in multiple of page sizes and granularity at which a double word gather operates is 4 bytes, thus there may not a chance of going beyond a page boundary while reading a double word from a normalized index which will always begin at byte address multiple of doubleword, @sviswa7's concern was around running into subsequent heap object header  (instance / array) in case gather array size is non power of two and a double word read access byte beyond sub-word array. Default ObjectAlignmentInBytes is 8 bytes and I expect it will never be less than 4 bytes as that is the word size for most of the targets. 

I am sitting over the performant solution for AVX512 targets.